cmd/restic/exclude.go

```diff
@@ -60,22 +68,43 @@ func rejectIfPresent(excludeFileSpec string) (RejectFunc, error) {
 	}
 	debug.Log("using %q as exclusion tagfile", tf)
 	fn := func(filename string, _ os.FileInfo) bool {
-		return isExcludedByFile(filename, tf, tc)
+		return isExcludedByFile(filename, tf, tc, rc)
 	}
 	return fn, nil
 }
 
 // isExcludedByFile interprets filename as a path and returns true if that file
 // is in a excluded directory. A directory is identified as excluded if it contains a
-// tagfile which bears the name specified in tagFilename and starts with header.
-func isExcludedByFile(filename, tagFilename, header string) bool {
+// tagfile which bears the name specified in tagFilename and starts with
+// header. If rc is non-nil, it is used to expedite the evaluation of a
+// directory based on previous visits.
+func isExcludedByFile(filename, tagFilename, header string, rc *rejectionCache) bool {
 	if tagFilename == "" {
 		return false
 	}
 	dir, base := filepath.Split(filename)
 	if base == tagFilename {
 		return false // do not exclude the tagfile itself
 	}
+	if rc != nil {
+		rc.mtx.Lock()
+		defer rc.mtx.Unlock()
+		if rejected, visited := rc.m[dir]; visited {
+			return rejected
+		}
+	}
+
+	rejected := isDirExcludedByFile(dir, tagFilename, header)
+	if rc != nil {
```

Comment:
This one too, e.g. `rc.Store(dir, rejected)`.

Reply:
Done.